From d66587b60e507b6d60fe5caff2c4511957df540b Mon Sep 17 00:00:00 2001
From: Jinping Wang <jinping.wang@amlogic.com>
Date: Tue, 20 Aug 2019 16:51:36 +0800
Subject: [PATCH] cec: support str wakeup by cec [1/3]

PD# SWPL-12320

Problem:
sometimes not go to wakeup source when resume from str

Solution:
support str wakeup by cec

Verify:
test ok using p321

Change-Id: Ibcfa83b3a378d11b6f63a8bd381e99e47f9f92ff
Signed-off-by: Jinping Wang <jinping.wang@amlogic.com>
---
 .../com/android/server/hdmi/HdmiCecLocalDeviceTv.java   | 14 ++++++++------
 .../com/android/server/hdmi/HdmiControlService.java     | 17 ++++++++++++++++-
 2 files changed, 24 insertions(+), 7 deletions(-)

diff --git a/services/core/java/com/android/server/hdmi/HdmiCecLocalDeviceTv.java b/services/core/java/com/android/server/hdmi/HdmiCecLocalDeviceTv.java
index b7a687e..1d040bd 100644
--- a/services/core/java/com/android/server/hdmi/HdmiCecLocalDeviceTv.java
+++ b/services/core/java/com/android/server/hdmi/HdmiCecLocalDeviceTv.java
@@ -161,7 +161,7 @@ final class HdmiCecLocalDeviceTv extends HdmiCecLocalDevice {
             HdmiDeviceInfo info = tvInfo.getHdmiDeviceInfo();
             if (info == null) return;
             addTvInput(inputId, info.getId());
-            if (info.isCecDevice()) {
+            if (info.isCecDevice() && mService.isInputChangeListenerReady()) {
                 processDelayedActiveSource(info.getLogicalAddress());
             }
         }
@@ -492,14 +492,16 @@ final class HdmiCecLocalDeviceTv extends HdmiCecLocalDevice {
                 HdmiLogger.debug("Device info %X not found; buffering the command", logicalAddress);
                 mDelayedMessageBuffer.add(message);
             }
-        } else if (isInputReady(info.getId())
-                || info.getDeviceType() == HdmiDeviceInfo.DEVICE_AUDIO_SYSTEM) {
+        } else if (!isInputReady(info.getId())) {
+            HdmiLogger.debug("Input not ready for device: %X; buffering the command", info.getId());
+            mDelayedMessageBuffer.add(message);
+        } else if (!mService.isInputChangeListenerReady()) {
+            HdmiLogger.debug("InputChangeListener not set for device: %X; buffering the command", info.getId());
+            mDelayedMessageBuffer.add(message);
+        } else {
             updateDevicePowerStatus(logicalAddress, HdmiControlManager.POWER_STATUS_ON);
             ActiveSource activeSource = ActiveSource.of(logicalAddress, physicalAddress);
             ActiveSourceHandler.create(this, null).process(activeSource, info.getDeviceType());
-        } else {
-            HdmiLogger.debug("Input not ready for device: %X; buffering the command", info.getId());
-            mDelayedMessageBuffer.add(message);
         }
         return true;
     }
diff --git a/services/core/java/com/android/server/hdmi/HdmiControlService.java b/services/core/java/com/android/server/hdmi/HdmiControlService.java
index e34977d..9f8c340 100755
--- a/services/core/java/com/android/server/hdmi/HdmiControlService.java
+++ b/services/core/java/com/android/server/hdmi/HdmiControlService.java
@@ -220,7 +220,7 @@ public class HdmiControlService extends SystemService {
             new ArrayList<>();
 
     @GuardedBy("mLock")
-    private InputChangeListenerRecord mInputChangeListenerRecord;
+    private InputChangeListenerRecord mInputChangeListenerRecord = null;
 
     @GuardedBy("mLock")
     private HdmiRecordListenerRecord mRecordListenerRecord;
@@ -1554,6 +1554,15 @@ public class HdmiControlService extends SystemService {
         public void setInputChangeListener(final IHdmiInputChangeListener listener) {
             enforceAccessPermission();
             HdmiControlService.this.setInputChangeListener(listener);
+            runOnServiceThread(new Runnable() {
+                @Override
+                public void run() {
+                    if (tv() != null) {
+                        tv().processAllDelayedMessages();
+                        HdmiLogger.debug("process All Delayed Message when set InputChangeListener");
+                    }
+                }
+            });
         }
 
         @Override
@@ -2023,6 +2032,12 @@ public class HdmiControlService extends SystemService {
         }
     }
 
+    public boolean isInputChangeListenerReady() {
+        synchronized (mLock) {
+            return (mInputChangeListenerRecord != null);
+        }
+    }
+
     private void removeInputChangeListener(IHdmiInputChangeListener listener) {
         synchronized (mLock) {
             Slog.d(TAG, "removeInputChangeListener");
-- 
2.10.2

